controllers/databaseclusterbackup_controller.go

```diff
@@ -221,10 +242,55 @@ func (r *DatabaseClusterBackupReconciler) reconcilePXC(ctx context.Context, back
 			Namespace: backup.Namespace,
 		},
 	}
+
+	// Given the more imperative-like nature of the DatabaseClusterBackup we
+	// can't afford to wait for the DatabaseCluster to be reconciled to have
+	// the necessary storage defined because this could lead to a failure and
+	// the backup wouldn't be retried. Instead, we check and update the
+	// storages map here to make sure the PerconaXtraDBCluster has the
+	// necessary storage defined to be able to perform the backup.
```

Comment:
Could this approach lead to race-conditions?
Such as a db cluster is being reconciled, but a backup is triggered and we overwrite dbcluster values or fail because of the revision number not matching?

Wouldn't it make more sense to keep this in the db cluster reconciliation loop and retry backup reconciliation until the storages are properly defined?

Reply:
> Wouldn't it make more sense to keep this in the db cluster reconciliation loop and retry backup reconciliation until the storages are properly defined?

Initially, I thought this wouldn't work because I was always thinking about immediately creating the PerconaXtraDBClusterBackup (before the DBC reconciliation had the opportunity to add the storages) and that would lead to a that backup request failing for ever, backup reconciliation wouldn't fix it.

Now I'm thinking that, if the storage isn't yet available, we can fail the backup reconciliation without creating the PerconaXtraDBClusterBackup. Then, the DBC reconciliation will add the storages and when the DBB reconcilation is run afterwards it will succeed and this should work.

I like this approach much better, I'll try to implement it and let you know. Thank you for questioning this and making me reevaluate my initial assumptions.

Reply:
We can introduce a new state here and use [ensureClusterIsReady](https://github.com/percona/everest-operator/blob/main/controllers/databaseclusterrestore_controller.go#L97-L115) that we have for restores